Photon power law

settwi opened the pull request:
PR

## PR Description
Adds photon space broken power law.

Compare with [f_bpow](https://ips.ucsd.edu/help/ssw/packages_f.html#F_BPOW) and [bknpower](https://heasarc.gsfc.nasa.gov/xanadu/xspec/manual/node142.html).

The broken power law bins flux properly and uses an analytical antiderivative of a power law to do so.
The documentation is (mostly?) in line with other stuff I've seen in sunxspex so far.
Needs more tests!

Fixes #101

settwi commented:
The Python 3.7 checks will always fail because I'm using the [walrus operator](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions)

DanRyanIrish reviewed (changes requested):
Hi @settwi.  This looks good.  Could you convert the "plot test" into an actual test where you check the output values?

settwi commented:
Yes! I'll work on that today. Thank you

KriSun95 commented on sunxspex/photon_power_law.py:
```diff
@@ -0,0 +1,100 @@
+    norm: PHOTON_RATE_UNIT,
+    index: u.one
+) -> (PHOTON_RATE_UNIT * u.keV):
+    """Evaluate the antiderivative of a power law at a given energy.
```
This looks great @settwi ! Can the docstrings be changed to be more similar to the other modules that roughly follow the Numpy style guide? E.g., 
```
"""
Description

Parameters
-----------
...
Returns
--------
...
"""```

KriSun95 commented on sunxspex/tests/test_photon_power_law.py:
```diff
@@ -0,0 +1,35 @@
+    flux = ppl.broken_power_law_binned_flux(**paramz)
+
+    fig, ax = plt.subplots(figsize=(8, 6))
+    ax.stairs(flux.value, edges.value)
```
This is likely my own ignorance here but why assign an `edges` variable and use `edges.value` rather than just use `paramz['energy_edges'].value`?

settwi replied:
No real reason. :)

samaloney commented on sunxspex/photon_power_law.py:
```diff
@@ -0,0 +1,100 @@
+    # norm is from continuity at break energy and solving.
+    norm = reference_flux * (reference_energy / break_energy)**(norm_idx)
+
+    cnd = energy_edges <= break_energy
+    lower = energy_edges[cnd]
+    upper = energy_edges[~cnd]
```
What if there is only one bin and one energy is below the break and the other is above? - I know it's another edge case

samaloney reviewed (changes requested):
Nice additions @settwi comments below are more of a discussion for us all have rather than you to address/fix

- Would a `power_law_binned_flux` be useful I know it can be achieved using the break energy but that's kinda opaque  
- Type annotations I'm a fan but we should probably make a decision as package/group
- Is the root folder the correct place for these kind of functions
- There are similar functions but for electrons in the thick target models could they be combined/integrated - less code is better than more code

settwi commented:
@samaloney thanks for the feedback! I'll address the style and testing feedback hopefully today and push up the changes.

settwi commented:
I squashed a couple bugs, adjusted the formatting, and tried to address the testing concerns that folks mentioned. Please take a look and let me know if other things need to change. Thanks!

PS: i am stuck trying to get Sphinx to properly build. weird indentation errors. ugh

@samaloney @DanRyanIrish @KriSun95

samaloney commented on sunxspex/photon_power_law.py:
```diff
@@ -0,0 +1,204 @@
+    r"""Evaluate the antiderivative of a power law at a given energy or vector of energies.
+
+    The power law antiderivative evaluated by this function is assumed to take the following form:
+    .. math:: f(E) = N \left( \frac{E}{E_0} \right)^{- \gamma},
```
I think this needs be either like this for inline 

```
:math:`f(E) = N \left( \frac{E}{E_0} \right)^{- \gamma}` where
```

or like this for block mode the white space and newlines are necessary

```
.. math:: 
   f(E) = N \left( \frac{E}{E_0} \right)^{- \gamma}

where 
```

You can see the rendered docs for this PR here https://sunxspex--103.org.readthedocs.build/en/103/api/sunxspex.photon_power_law.broken_power_law_binned_flux.html#sunxspex.photon_power_law.broken_power_law_binned_flux

samaloney commented on changelog/103.feature.rst:
```diff
@@ -0,0 +1 @@
+Add photon space broken power law, located in `sunxspex/photon_power_law.py`, function name `broken_power_law_binned_flux`. Behavior is equivalent to the OSPEX power law when the pivot option is given. Similar to XSPEC bknpower. See the pull request for more details.
```
```suggestion
Add photon space broken power law :func:`sunxspex.photon_power_law.broken_power_law_binned_flux`. Behavior is equivalent to the OSPEX power law when the pivot option is given. Similar to XSPEC bknpower.
```

samaloney commented on sunxspex/photon_power_law.py:
```diff
@@ -0,0 +1,204 @@
+    reference_flux: PHOTON_RATE_UNIT,
+    index: u.one
+) -> PHOTON_RATE_UNIT:
+    '''Single power law, defined by setting the break energy to -inf and the lower index to nan.
```
```suggestion
    r'''Single power law, defined by setting the break energy to -inf and the lower index to nan.
```

samaloney commented:
No worries a fresh install in a clean virtual env e.g. `pip install -e .[dev]` and it should _just work_ but can be finicky.

As far as code is concerned I think the PR is good to go but I think there some 'meta' issues that need to be discussed I had suggest at the proposed monthly call?

KriSun95 commented:
I think this PR might be ready to merge. Should we add it in?

settwi commented:
@samaloney @DanRyanIrish could one of you approve the PR if you think it's good to go?

DanRyanIrish commented on sunxspex/tests/test_photon_power_law.py:
```diff
@@ -0,0 +1,144 @@
+    npt.assert_allclose(flux1, flux2)
+
+
+if __name__ == '__main__':
+    def all_tests():
+        test_integration()
+        test_different_bins()
+        test_empty_or_single_energy_edges()
+        test_no_flux()
+        test_one_big_bin()
+        test_single_power_law()
+
+    all_tests()
```
Is this necessary?  Won't that all be run by the CI without this?

settwi replied:
Yes, I wanted to run the tests locally

DanRyanIrish commented on sunxspex/photon_power_law.py:
```diff
@@ -0,0 +1,206 @@
+import functools
+
+import numpy as np
+
+import astropy.units as u
```
```suggestion
import astropy.units as u
import numpy as np
```
See section on ordering imports in PEP 8

settwi replied:
I actually ran `autopep8` on this file multiple times :p

It looks like [this example given in the PEP8 "Imports" section](https://peps.python.org/pep-0008/#imports) shows that the `foo.bar` package should be imported second?
```py
import myclass
import foo.bar.yourclass
```

DanRyanIrish reviewed (commented):
This looks OK to me.  A couple general style points though.

I think it's good practice to have readable variable names.  In the tests and main code there are several 2- and 3-letter variable names that are not descriptive, making the code harder to understand.

Similarly, it's good practice name functions with verbs.  That way you can tell it's a function/method from its name., e.g. `power_law_binned_flux` -> `calculate_power_law_binned_flux`

settwi commented:
Sure, I can update the variables.On Jul 12, 2023, at 16:50, DanRyanIrish ***@***.***> wrote:﻿
@DanRyanIrish commented on this pull request.

This looks OK to me.  A couple general style points though.
I think it's good practice to have readable variable names.  In the tests and main code there are several 2- and 3-letter variable names that are not descriptive, making the code harder to understand.
Similarly, it's good practice name functions with verbs.  That way you can tell it's a function/method from its name., e.g. power_law_binned_flux -> calculate_power_law_binned_flux

In sunxspex/tests/test_photon_power_law.py:
> +if __name__ == '__main__':
+    def all_tests():
+        test_integration()
+        test_different_bins()
+        test_empty_or_single_energy_edges()
+        test_no_flux()
+        test_one_big_bin()
+        test_single_power_law()
+
+    all_tests()

Is this necessary?  Won't that all be run by the CI without this?



In sunxspex/photon_power_law.py:
> +import numpy as np
+
+import astropy.units as u

⬇️ Suggested change
-import numpy as np
-
-import astropy.units as u
+import astropy.units as u
+import numpy as np


See section on ordering imports in PEP 8

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

hayesla commented on sunxspex/photon_power_law.py:
```diff
@@ -0,0 +1,204 @@
+def compute_broken_power_law(
+    energy_edges: u.keV,
+    reference_energy: u.keV,
+    reference_flux: PHOTON_RATE_UNIT,
```
if there a reason this is called "reference_flux" rather that "normalization"?

settwi replied:
i think it's an artifact from when the function had a slightly different implementation. i'll update the name!

hayesla commented on sunxspex/photon_power_law.py:
```diff
@@ -0,0 +1,204 @@
+    reference_energy : `astropy.units.Quantity`
+        Energy at which the normalization is applied, i.e. :math:`E_0`.
+    reference_flux : `astropy.units.Quantity`
+        Normalization flux for the photon power law.
```
I would also add another description here for what this normalisation is

hayesla reviewed (approved):
this looks great @settwi!

DanRyanIrish reviewed (commented):
Hi @settwi.  Sorry for taking so long to look at this.  It also also looks good to me, although I notice a couple of the tests are failing.  I've set the tests rerunning.  Once they are passing and @hayesla's comments are addressed, we should get this merged!

settwi commented:
hey @DanRyanIrish @hayesla i will implement Laura's comments. thanks for checking in.
as for the tests: i think some of them are not my fault that they don't pass. for one of them there is something messed up on the CI side i believe. the ones that i can fix i will fix though.
thanks!

settwi commented:
@hayesla and @DanRyanIrish i've implemented Laura's comments and i think things are good to go. i know some tests are failing but i think most of them are out of my hands...

settwi commented:
@samaloney @DanRyanIrish could you mark the pending changes as approved?

KriSun95 commented:
@settwi reminded me of a conversation with @DanRyanIrish at SPD that this was OK to be accepted from his point of view so, before any larger changes are made to the repo, I'm going to merge this PR. Any other changes down the line can be addressed in the refactoring.
